File: src/Range.cpp
// Range.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// Range.C

#include "Range.H"

Range::Range() {
  empty_ = true;
}

Range::Range(double x) {
  min_ = max_ = x;
  empty_ = false;
}

Range::Range(double x, double y) {
  min_ = max_ = x;
  empty_ = false;
  extend(y);
}

bool Range::empty() const {
  return empty_;
}

void Range::extend(double x) {
  if (empty_) {
    min_ = max_ = x;
    empty_ = false;
  } else {
    if (x<min_)
      min_=x;
    if (x>max_)
      max_=x;
  }
}

void Range::unionize(Range const &r) {
  if (r.empty_)
    return;
  if (empty_) {
    min_ = r.min_;
    max_ = r.max_;
    empty_ = false;
  } else {
    if (r.min_<min_)
      min_ = r.min_;
    if (r.max_>max_)
      max_ = r.max_;
  }
}

void Range::intersect(Range const &r) {
  if (empty_)
    return;
  if (r.empty_) {
    empty_ = true;
  } else {
    if (r.min_>min_)
      min_ = r.min_;
    if (r.max_<max_)
      max_ = r.max_;
    if (min_>max_)
      empty_ = true;
  }
}

bool Range::contains(double x) const {
  if (empty_)
    return false;
  else
    return x>=min_ && x<=max_;
}

double Range::range() const {
  if (empty_)
    return 0;
  else
    return max_ - min_;
}

File: python/qplot_backend/range_.py
# range.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems

import math


class Range:
    """Tracks the min/max extent of a set of values.
    An empty Range has no values added yet."""

    def __init__(self, x: float = None, y: float = None):
        if x is None:
            self._empty = True
            self._min = self._max = 0.0
        else:
            self._empty = False
            self._min = self._max = x
            if y is not None:
                self.extend(y)

    def empty(self) -> bool:
        return self._empty

    def min(self) -> float:
        return self._min

    def max(self) -> float:
        return self._max

    def extend(self, x: float) -> None:
        """Expand the range to include x. NaN values are ignored."""
        if math.isnan(x):
            return
        if self._empty:
            self._min = self._max = x
            self._empty = False
        elif x < self._min:
            self._min = x
        elif x > self._max:
            self._max = x

    def unionize(self, other: "Range") -> None:
        """Expand this range to also cover other."""
        if other._empty:
            return
        if self._empty:
            self._min = other._min
            self._max = other._max
            self._empty = False
        else:
            if other._min < self._min:
                self._min = other._min
            if other._max > self._max:
                self._max = other._max

    def intersect(self, other: "Range") -> None:
        """Shrink this range to only the overlap with other."""
        if self._empty:
            return
        if other._empty:
            self._empty = True
        else:
            if other._min > self._min:
                self._min = other._min
            if other._max < self._max:
                self._max = other._max
            if self._min > self._max:
                self._empty = True

    def contains(self, x: float) -> bool:
        if self._empty:
            return False
        return self._min <= x <= self._max

    def range(self) -> float:
        """Returns the span (max - min), or 0 for an empty range."""
        if self._empty:
            return 0.0
        return self._max - self._min
